src/charmed_kubeflow_chisme/testing/_unit_tests.py

```diff
@@ -0,0 +1,68 @@
+# Copyright 2022 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+from contextlib import nullcontext as does_not_raise
+
+import pytest
+from oci_image import MissingResourceError
+from ops.model import BlockedStatus, WaitingStatus
+
+
+def test_leadership_events(harness):
+    """Test leader-elected event handling."""
+    harness.set_leader(False)
+    harness.begin_with_initial_hooks()
+    assert harness.charm.model.unit.status == WaitingStatus("Waiting for leadership")
+    harness.set_leader(True)
+    assert harness.charm.model.unit.status != WaitingStatus("Waiting for leadership")
+    harness.set_leader(False)
+    # Emit another leader_elected event due to https://github.com/canonical/operator/issues/812
+    harness._charm.on.leader_elected.emit()
+    assert harness.charm.model.unit.status == WaitingStatus("Waiting for leadership")
+
+
+def test_missing_relation(harness):
+    """Test if status is blocked when relation is missing."""
+    harness.set_leader(True)
+    harness.add_oci_resource(
+        "oci-image",
+        {
+            "registrypath": "image",
+            "username": "",
+            "password": "",
+        },
+    )
+    harness.begin_with_initial_hooks()
+    assert isinstance(harness.charm.model.unit.status, WaitingStatus)
+
+
+def test_missing_image(harness):
+    """Test if status is blocked when image is missing."""
+    harness.set_leader(True)
+    harness.begin_with_initial_hooks()
+    assert isinstance(harness.charm.model.unit.status, BlockedStatus)
```

Comment:
You are right about what `test_missing_relation` is meant for. I think it's a common standard in kubeflow charms to first check image presence and then the relations, which is the reason for this test structure. Most of our charms have oci image.
I parametrized status for both `test_missing_image` and `test_missing_relation` since I noticed it may be either `BlockedStatus` or `WaitingStatus`.
I'm not sure if `relation_name` is needed - it doesn't matter what other charm(s) it has to be related to, the purpose is to check the status after coming across the first missing relation.
I added a readme which should help to understand these tests better. Let me know what you think!

Reply:
Yeah I see what you're saying.  It seems like these tests then work well, but only if the charm author follows a particular convention (specific test order, etc).  I worry that someone looking at these tests won't realise the common standard you mentioned.  

Ideally, we would not have the restriction at all, but that would require some changes to how we handle things like mentioned in the previous comment...  Is there a good way to document what users need to do (maybe show the charm structure expected in order to use these tests?)

Reply:
I'm not sure how to tackle this part. Those tests are made to meet the order of checks applied in most of our charms. Do you think the readme makes things more clear?